### PR TITLE
feat: Verify provenance by build type

### LINKED ISSUE #473: [cleanup] Use a static mapping for supported provenance versions
Instead of `init` methods, which are fragile.

### LINKED ISSUE #599: [feature][byob] verification for sha1 provided by TRW
verification counterpart for https://github.com/slsa-framework/slsa-github-generator/issues/2079

To think about: if the sha1 != GitHub event, options like `--source-branch`, `--source-tag` or another source option should fail because we can no longer trust the event?

For generators, I think it's fine to trust keep the code as-is because the options refer to the trigger workflow. For builders, I think we need to fail.

### LINKED ISSUE #600: [feature][byob] Verification for sha1 provided by TRW for v1.0
verification counterpart for https://github.com/slsa-framework/slsa-github-generator/issues/2137

Make sure to follow the comments in https://github.com/slsa-framework/slsa-verifier/issues/599 and differentiate between generators and builders

### PR DESCRIPTION
**Pull request by ianlewis**
Fixes #473

Updates handling of provenance by providing implementations based on [buildType](https://slsa.dev/provenance/v1#buildType) since this determines how to interpret parameters and dependencies.  This is done because we need a way to interpret parameters not just based on the predicateType. The 3 major build types with format differences are:
  - non-BYOB SLSA v0.2
  - BYOB SLSA v0.2
  - BYOB SLSA v1.0

### CODE REVIEW

File: verifiers/internal/gha/slsaprovenance/v1.0/byob.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+	return p.prov.Predicate
+}
+
+// BuilderID implements Provenance.BuilderID.
+func (p *BYOBProvenanceV1) BuilderID() (string, error) {
+	return p.prov.Predicate.RunDetails.Builder.ID, nil
+}
+
+// SourceURI implements Provenance.SourceURI.
+func (p *BYOBProvenanceV1) SourceURI() (string, error) {
```
Comment (ianlewis) [inline verifiers/internal/gha/slsaprovenance/v1.0/byob.go:40]: This method was reverted. Need to revert to the original.

File: verifiers/internal/gha/slsaprovenance/v0.2/base.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+func (p *provenanceV02) Subjects() ([]intoto.Subject, error) {
+	subj := p.prov.Subject
+	if len(subj) == 0 {
+		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, "no subjects")
+	}
+	return subj, nil
+}
+
+// GetBranch implements Provenance.GetBranch.
+func (p *provenanceV02) GetBranch() (string, error) {
```
Comment (ianlewis) [inline verifiers/internal/gha/slsaprovenance/v0.2/base.go:63]: Need to support when `sha1` is passed

File: verifiers/internal/gha/slsaprovenance/v0.2/base.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	environment, ok := p.prov.Predicate.Invocation.Environment.(map[string]interface{})
+	if !ok {
+		return "", fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, "parameters type")
+	}
+
+	return common.GetBranch(environment, p.upperEnv)
+}
+
+// GetTag implements Provenance.GetTag.
+func (p *provenanceV02) GetTag() (string, error) {
```
Comment (ianlewis) [inline verifiers/internal/gha/slsaprovenance/v0.2/base.go:74]: Need to support when `sha1` is passed.

**Comment by ianlewis** [comment]
Blocked on https://github.com/slsa-framework/github-actions-buildtypes/issues/4

**Comment by ianlewis** [comment]
This should support BYOB TRW generated binaries w/ `verify-artifact`.

File: verifiers/internal/gha/slsaprovenance/v1.0/byob.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+	"github.com/slsa-framework/slsa-verifier/v2/verifiers/internal/gha/slsaprovenance/common"
+)
+
+var (
+	// byobBuildType is the base build type for BYOB delegated builders.
+	byobBuildType = "https://github.com/slsa-framework/slsa-github-generator/delegator-generic@v0"
+
+	// containerBasedBuildType is the build type for the container-based builder and is based on BYOB.
+	containerBasedBuildType = "https://slsa.dev/container-based-build/v0.1?draft"
```
Comment (laurentsimon) [inline verifiers/internal/gha/slsaprovenance/v1.0/byob.go:20]: note: the container-based builder is not based no BYOB, iiuc https://github.com/slsa-framework/slsa-github-generator/blob/main/.github/workflows/builder_container-based_slsa3.yml

**Review by laurentsimon — APPROVED** [review]
Thanks.

PTAL at the container-based part, which is not based on BYOB

File: verifiers/internal/gha/slsaprovenance/v1.0/byob.go (hunk 4 above)
Comment (ianlewis) [reply verifiers/internal/gha/slsaprovenance/v1.0/byob.go:20]: Right, I'll refactor this a bit.

**Pull request merged**